blueoil/utils/smartdict.py

```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 The Blueoil Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+
+class SmartDict(dict):
+    def __init__(self, d=None, **kwargs):
+        super(SmartDict, self).__init__()
+        self.update(d, **kwargs)
+
+    def update(self, d=None, **kwargs):
+        d = d or {}
+        d.update(kwargs)
+        for key, value in d.items():
+            self[key] = value
+
+    def __setitem__(self, name, value):
+        if isinstance(value, (list, tuple)):
+            value = [
+                self.__class__(x) if isinstance(x, dict) else x
+                for x in value
+            ]
+        elif isinstance(value, dict) and not isinstance(value, self.__class__):
+            value = self.__class__(value)
+        super(SmartDict, self).__setitem__(name, value)
+
+    def __getattr__(self, name):
+        if name in self:
+            return self[name]
+        return super(SmartDict, self).__getattr__(name)
```

Comment:
I think `__getattr__` should raise an AttributeError instead of delegating to the super class. Because Dict doesn't have a method `__getattr__`, the current error message is like:
```
AttributeError: 'super' object has no attribute '__getattr__'
```
It should be like:
```
AttributeError: 'SmartDict' object has no attribute 'x'
```